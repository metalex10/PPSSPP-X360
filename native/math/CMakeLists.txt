set(SRCS
  lin/matrix4x4.cpp
  lin/vec3.cpp
  lin/quat.cpp
  lin/aabb.cpp
	curves.cpp
	math_util.cpp
  )

set(SRCS ${SRCS})

add_library(lin STATIC ${SRCS})

if(UNIX)
  add_definitions(-fPIC)
endif(UNIX)
