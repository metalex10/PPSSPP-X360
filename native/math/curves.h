#pragma once

#include <base/basictypes.h>

// Easy curve computation for fades etc.


// output range: [0.0, 1.0]
float linearInOut(int t, int fadeInLength, int solidLength, int fadeOutLength);
float linearIn(int t, int fadeInLength);
float linearOut(int t, int fadeInLength);

// smooth operator [0, 1] -> [0, 1]
float ease(float val);
float ease(int t, int fadeLength);

// need a bouncy ease

// waveforms [0, 1]
float sawtooth(int t, int period);

// output range: -1.0 to 1.0
float passWithPause(int t, int fadeInLength, int pauseLength, int fadeOutLength);
