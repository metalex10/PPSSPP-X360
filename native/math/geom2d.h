#pragma once

#include <cmath>

struct Point {
	Point() {}
	Point(float x_, float y_) : x(x_), y(y_) {}

	float x;
	float y;

	float distanceTo(const Point &other) const {
		float dx = other.x - x, dy = other.y - y;
		return sqrtf(dx*dx + dy*dy);
	}

	/*
	FocusDirection directionTo(const Point &other) const {
		int angle = atan2f(other.y - y, other.x - x) / (2 * M_PI) - 0.125;

	}*/
};


// Resolved bounds on screen after layout.
struct Bounds {
	Bounds() : x(0), y(0), w(0), h(0) {}
	Bounds(float x_, float y_, float w_, float h_) : x(x_), y(y_), w(w_), h(h_) {}

	bool Contains(float px, float py) const {
		return (px >= x && py >= y && px < x + w && py < y + h);
	}

	bool Intersects(const Bounds &other) const {
		return !(x > other.x2() || x2() < other.x || y > other.y2() || y2() < other.y);
	}

	void Clip(const Bounds &clipTo) {
		if (x < clipTo.x) {
			w -= clipTo.x - x;
			x = clipTo.x;
		}
		if (y < clipTo.y) {
			h -= clipTo.y - y;
			y = clipTo.y;
		}
		if (x2() > clipTo.x2()) {
			w = clipTo.x2() - x;
		}
		if (y2() > clipTo.y2()) {
			h = clipTo.y2() - y;
		}
	}

	float x2() const { return x + w; }
	float y2() const { return y + h; }
	float centerX() const { return x + w * 0.5f; }
	float centerY() const { return y + h * 0.5f; }
	Point Center() const {
		return Point(centerX(), centerY());
	}

	float x;
	float y;
	float w;
	float h;
};


