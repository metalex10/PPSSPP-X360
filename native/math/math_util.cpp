#include "math/math_util.h"
#include <stdlib.h>

#if defined(__ARM_ARCH_7A__) && !defined(BLACKBERRY)

void EnableFZ()
{
	int x;
	asm(
		"fmrx %[result],FPSCR \r\n"
		"orr %[result],%[result],#16777216 \r\n"
		"fmxr FPSCR,%[result]"
		:[result] "=r" (x) : :
	);
	//printf("ARM FPSCR: %08x\n",x);
}
#else
void EnableFZ()
{
	// TODO
}
#endif
