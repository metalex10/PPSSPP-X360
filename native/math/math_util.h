#ifndef _MATHUTILS_H
#define _MATHUTILS_H

#include <cmath>
#include <cstring>

#include "base/basictypes.h"

inline float sqr(float f)	{return f*f;}
inline float sqr_signed(float f) {return f<0 ? -f*f : f*f;}

typedef unsigned short float16;

// This ain't a 1.5.10 float16, it's a stupid hack format where we chop 16 bits off a float.
// This choice is subject to change. Don't think I'm using this for anything at all now anyway.
// DEPRECATED
inline float16 FloatToFloat16(float x) {
	int ix;
	memcpy(&ix, &x, sizeof(float));
	return ix >> 16;
}

inline float Float16ToFloat(float16 ix) {
	float x;
	memcpy(&x, &ix, sizeof(float));
	return x;
}

inline uint32_t RoundUpToPowerOf2(uint32_t v) {
	v--;
	v |= v >> 1;
	v |= v >> 2;
	v |= v >> 4;
	v |= v >> 8;
	v |= v >> 16;
	v++;
	return v;
}

#define PI 3.141592653589793f
#ifndef M_PI
#define M_PI 3.141592653589793f
#endif

// The stuff in this file is from all over the web, esp. dspmusic.org. I think it's all public domain.
// In any case, very little of it is used anywhere at the moment.

// Calculate pseudo-random 32 bit number based on linear congruential method. 
void SetSeed(unsigned int seed);
unsigned int GenerateRandomNumber();
inline float GenerateRandomFloat01() {
	return (float)((double)GenerateRandomNumber() / 0xFFFFFFFF);
}
inline float GenerateRandomSignedFloat() {
	return (float)((double)GenerateRandomNumber() / 0x80000000) - 1.0f;
}


inline float GaussRand()
{
	float R1 = GenerateRandomFloat01();
	float R2 = GenerateRandomFloat01();

	float X = sqrtf(-2.0f * logf(R1)) * cosf(2.0f * PI * R2);
	if (X > 4.0f) X = 4.0f;
	if (X < -4.0f) X = -4.0f;
	return X;
}

// Accuracy unknown
inline double atan_fast(double x) {
	return (x / (1.0 + 0.28 * (x * x)));
}


// linear -> dB conversion
inline float lin2dB(float lin) {
	const float LOG_2_DB = 8.6858896380650365530225783783321f;	// 20 / ln( 10 )
	return logf(lin) * LOG_2_DB;
}

// dB -> linear conversion
inline float dB2lin(float dB) {
	const float DB_2_LOG = 0.11512925464970228420089957273422f;	// ln( 10 ) / 20
	return expf(dB * DB_2_LOG);
}

inline bool my_isinf(float f) {
	union {
		float f;
		uint32_t u;
	} f2u;
	f2u.f = f;
	return f2u.u == 0x7f800000 ||
		f2u.u == 0xff800000;
}

inline bool my_isnan(float f) {
	union {
		float f;
		uint32_t u;
	} f2u;
	f2u.f = f;
	// NaNs have non-zero mantissa
	return ((f2u.u & 0x7F800000) == 0x7F800000) && (f2u.u & 0x7FFFFF);
}

inline bool my_isnanorinf(float f) {
	union {
		float f;
		uint32_t u;
	} f2u;
	f2u.f = f;
	// NaNs have non-zero mantissa, infs have zero mantissa. That is, we just ignore the mantissa here.
	return ((f2u.u & 0x7F800000) == 0x7F800000);
}

inline int is_even(float d) {
	float int_part;
	modff(d / 2.0f, &int_part);
	return 2.0f * int_part == d;
}

// Rounds *.5 to closest even number
inline double round_ieee_754(double d) {
	float i = floor(d);
	d -= i;
	if (d < 0.5f)
		return i;
	if (d > 0.5f)
		return i + 1.0f;
	if (is_even(i))
		return i;
	return i + 1.0f;
}

union FP32 {
	uint32_t u;
	float f;
};

struct FP16 {
	uint16_t u;
};

// magic code from ryg: http://fgiesen.wordpress.com/2012/03/28/half-to-float-done-quic/
// See also SSE2 version: https://gist.github.com/rygorous/2144712
inline FP32 half_to_float_fast5(FP16 h)
{
	static const FP32 magic = { (127 + (127 - 15)) << 23 };
	static const FP32 was_infnan = { (127 + 16) << 23 };
	FP32 o;
	o.u = (h.u & 0x7fff) << 13;     // exponent/mantissa bits
	o.f *= magic.f;                 // exponent adjust
	if (o.f >= was_infnan.f)        // make sure Inf/NaN survive
		o.u |= 255 << 23;
	o.u |= (h.u & 0x8000) << 16;    // sign bit
	return o;
}

inline float ExpandHalf(uint16_t half) {
	FP16 fp16;
	fp16.u = half;
	FP32 fp = half_to_float_fast5(fp16);
	return fp.f;
}

// More magic code: https://gist.github.com/rygorous/2156668
inline FP16 float_to_half_fast3(FP32 f)
{
	static const FP32 f32infty = { 255 << 23 };
	static const FP32 f16infty = { 31 << 23 };
	static const FP32 magic = { 15 << 23 };
	static const uint32_t sign_mask = 0x80000000u;
	static const uint32_t round_mask = ~0xfffu;
	FP16 o = { 0 };

	uint32_t sign = f.u & sign_mask;
	f.u ^= sign;

	if (f.u >= f32infty.u) // Inf or NaN (all exponent bits set)
		o.u = (f.u > f32infty.u) ? 0x7e00 : 0x7c00; // NaN->qNaN and Inf->Inf
	else // (De)normalized number or zero
	{
		f.u &= round_mask;
		f.f *= magic.f;
		f.u -= round_mask;
		if (f.u > f16infty.u) f.u = f16infty.u; // Clamp to signed infinity if overflowed

		o.u = f.u >> 13; // Take the bits!
	}

	o.u |= sign >> 16;
	return o;
}

inline uint16_t ShrinkToHalf(float full) {
	FP32 fp32;
	fp32.f = full;
	FP16 fp = float_to_half_fast3(fp32);
	return fp.u;
}

// FPU control.
void EnableFZ();

#endif
