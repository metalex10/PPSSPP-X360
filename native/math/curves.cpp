
#include <math.h>

#include "math/math_util.h"
#include "curves.h"

float linearInOut(int t, int fadeInLength, int solidLength, int fadeOutLength) {
	if (t < 0) return 0;
	if (t < fadeInLength) {
		return (float)t / fadeInLength;
	}
	t -= fadeInLength;
	if (t < solidLength) {
		return 1.0f;
	}
	t -= solidLength;
	if (t < fadeOutLength) {
		return 1.0f - (float)t / fadeOutLength;
	}
	return 0.0f;
}

float linearIn(int t, int fadeInLength) {
	if (t < 0) return 0;
	if (t < fadeInLength) {
		return (float)t / fadeInLength;
	}
	return 1.0f;
}

float linearOut(int t, int fadeOutLength) {
	return 1.0f - linearIn(t, fadeOutLength);
}

float ease(float val) {
	if (val > 1.0f) return 1.0f;
	if (val < 0.0f) return 0.0f;
	return ((-cosf(val * PI)) + 1.0f) * 0.5;
}

float ease(int t, int fadeLength)
{
	if (t < 0) return 0.0f;
	if (t >= fadeLength) return 1.0f;
	return ease((float)t / (float)fadeLength);
}

float sawtooth(int t, int period) {
	return (t % period) * (1.0f / (period - 1));
}

float passWithPause(int t, int fadeInLength, int pauseLength, int fadeOutLength)
{
	if (t < fadeInLength) {
		return -1.0f + (float)t / fadeInLength;
	}
	t -= fadeInLength;
	if (t < pauseLength) {
		return 0.0f;
	}
	t -= pauseLength;
	if (t < fadeOutLength) {
		return (float)t / fadeOutLength;
	}
	return 1.0f;
}
