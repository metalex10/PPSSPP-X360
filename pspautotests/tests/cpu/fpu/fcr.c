#include <common.h>
#include <math.h>
#include <float.h>
#include <limits.h>

void __attribute__((noinline)) printfcrs(const char *title) {
	unsigned int result[6] = {-1, -1, -1, -1, -1, -1};

	asm volatile (
		".set    noreorder\n"

		"cfc1    %0, $0\n"
		"cfc1    %1, $25\n"
		"cfc1    %2, $26\n"
		"cfc1    %3, $27\n"
		"cfc1    %4, $28\n"
		"cfc1    %5, $31\n"

		: "+r"(result[0]), "+r"(result[1]), "+r"(result[2]), "+r"(result[3]), "+r"(result[4]), "+r"(result[5])
	);

	printf("%s:\n  fcr0: %08x, fcr25: %08x, fcr26: %08x, fcr27: %08x, fcr28: %08x, fcr31: %08x\n", title, result[0], result[1], result[2], result[3], result[4], result[5]);
}

void __attribute__((noinline)) setfcr0(u32 v) {
	asm volatile (
		".set    noreorder\n"

		"ctc1    %0, $0\n"

		:
		: "r"(v)
	);
}

void __attribute__((noinline)) updatefcr31(u32 v) {
	asm volatile (
		".set    noreorder\n"

		"or      $v0, $0, %0\n"
		"ctc1    $v0, $31\n"
		"nop\n"

		:
		: "r"(v)
		: "v0"
	);
}

void __attribute__((noinline)) sqrtneg() {
	asm volatile (
		".set    noreorder\n"

		"lui     $v0, 0xBF80\n" // 0xBF800000 = -1
		"mtc1    $v0, $f0\n"
		"sqrt.s  $f0, $f0\n"

		:
		:
		: "v0", "$f0"
	);
}

int main(int argc, char *argv[]) {
	printfcrs("Initial");

	setfcr0(0xDEADBEEF);
	printfcrs("Update fcr0");
	updatefcr31(0x00000003);
	printfcrs("Update rounding mode (RM)");
	updatefcr31(0x0000007C);
	printfcrs("Update flags");
	updatefcr31(0x00000F80);
	printfcrs("Update enables");
	updatefcr31(0x0001F000);
	printfcrs("Update cause");
	updatefcr31(0x0001F000);
	// Crashes.
	//updatefcr31(0x00020000);
	//printfcrs("Update unimplemented (E)");
	updatefcr31(0x01000000);
	printfcrs("Update flushing (FS)");
	updatefcr31(0x00400000);
	printfcrs("Update flushing (FO)");
	updatefcr31(0x00200000);
	printfcrs("Update flushing (FN)");
	updatefcr31(0x00800000);
	printfcrs("Update FCC");
	updatefcr31(0xFE000000);
	printfcrs("Update FCC1-7");
	updatefcr31(0x001C0000);
	printfcrs("Update unknown");

	printf("\nFlag situations:\n");

	float f1;
	float f2;
	float f3;

	updatefcr31(0x00000000);
	sqrtneg();
	printfcrs("sqrt(-1)");

	updatefcr31(0x00000000);
	f1 = 0.0;
	f2 = 0.0;
	f3 = f1 / f2;
	printfcrs("Divide by zero");

	updatefcr31(0x00000000);
	f1 = NAN;
	f2 = NAN;
	f3 = f1 * f2;
	printfcrs("NAN math");

	updatefcr31(0x00000000);
	f1 = FLT_MAX;
	f2 = FLT_MAX;
	f3 = f1 * f2;
	printfcrs("Overflow");

	updatefcr31(0x00000000);
	f1 = 1.0;
	f2 = FLT_MAX;
	f3 = f1 / f2;
	printfcrs("Underflow");

	updatefcr31(0x00000000);
	f1 = 1.0;
	f2 = 3.0;
	f3 = f1 / f2;
	printfcrs("Inexact");

	return 0;
}