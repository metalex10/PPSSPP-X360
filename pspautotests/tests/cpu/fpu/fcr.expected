Initial:
  fcr0: 00003351, fcr25: 00000000, fcr26: 00000000, fcr27: 00000000, fcr28: 00000000, fcr31: 00000e00
Update fcr0:
  fcr0: 00003351, fcr25: 00000000, fcr26: 00000000, fcr27: 00000000, fcr28: 00000000, fcr31: 00000e00
Update rounding mode (RM):
  fcr0: 00003351, fcr25: 00000000, fcr26: 00000000, fcr27: 00000000, fcr28: 00000000, fcr31: 00000003
Update flags:
  fcr0: 00003351, fcr25: 00000000, fcr26: 00000000, fcr27: 00000000, fcr28: 00000000, fcr31: 0000007c
Update enables:
  fcr0: 00003351, fcr25: 00000000, fcr26: 00000000, fcr27: 00000000, fcr28: 00000000, fcr31: 00000f80
Update cause:
  fcr0: 00003351, fcr25: 00000000, fcr26: 00000000, fcr27: 00000000, fcr28: 00000000, fcr31: 0001f000
Update flushing (FS):
  fcr0: 00003351, fcr25: 00000000, fcr26: 00000000, fcr27: 00000000, fcr28: 00000000, fcr31: 01000000
Update flushing (FO):
  fcr0: 00003351, fcr25: 00000000, fcr26: 00000000, fcr27: 00000000, fcr28: 00000000, fcr31: 00000000
Update flushing (FN):
  fcr0: 00003351, fcr25: 00000000, fcr26: 00000000, fcr27: 00000000, fcr28: 00000000, fcr31: 00000000
Update FCC:
  fcr0: 00003351, fcr25: 00000000, fcr26: 00000000, fcr27: 00000000, fcr28: 00000000, fcr31: 00800000
Update FCC1-7:
  fcr0: 00003351, fcr25: 00000000, fcr26: 00000000, fcr27: 00000000, fcr28: 00000000, fcr31: 00000000
Update unknown:
  fcr0: 00003351, fcr25: 00000000, fcr26: 00000000, fcr27: 00000000, fcr28: 00000000, fcr31: 00000000

Flag situations:
sqrt(-1):
  fcr0: 00003351, fcr25: 00000000, fcr26: 00000000, fcr27: 00000000, fcr28: 00000000, fcr31: 00010040
Divide by zero:
  fcr0: 00003351, fcr25: 00000000, fcr26: 00000000, fcr27: 00000000, fcr28: 00000000, fcr31: 00010040
NAN math:
  fcr0: 00003351, fcr25: 00000000, fcr26: 00000000, fcr27: 00000000, fcr28: 00000000, fcr31: 00010040
Overflow:
  fcr0: 00003351, fcr25: 00000000, fcr26: 00000000, fcr27: 00000000, fcr28: 00000000, fcr31: 00005014
Underflow:
  fcr0: 00003351, fcr25: 00000000, fcr26: 00000000, fcr27: 00000000, fcr28: 00000000, fcr31: 0000300c
Inexact:
  fcr0: 00003351, fcr25: 00000000, fcr26: 00000000, fcr27: 00000000, fcr28: 00000000, fcr31: 00001004
